### PR TITLE
HEEDLS-281 Fix tutorial IDs on diagnostic page

### PR DESCRIPTION
**Pull request by benjimarshall**
## Changes
Change way tutorial IDs are selected for diagnostic page, and update
criteria for choosing diagnostic tutorials.

## Testing
Rewrite and add tests for diagnostic assessment service. I've gone through
everything and I'm reasonably confident that I've found all of the edge cases now.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs [hunk 1]
```diff
@@ -178,12 +153,25 @@ public void Get_diagnostic_assessment_should_return_null_if_archived_date_is_not
...
+        [Test]
+        public void Get_diagnostic_assessment_should_return_null_if_all_diagStatus_are_false()
         {
             // Given
-            const int customisationId = 1530;
-            const int candidateId = 23573;
-            const int sectionId = 74;
+            const int customisationId = 4831;
+            const int candidateId = 11;
+            const int sectionId = 75;
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs:174]: Why did we need to change this data? 🙂

File: DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs [hunk 2]
```diff
@@ -192,61 +180,166 @@ public void Get_diagnostic_assessment_should_return_null_if_diag_status_and_stat
...
+
+            // Then
+            result.Should().NotBeNull();
+            var tutorialIds = result!.Tutorials.Select(tutorial => tutorial.Id).ToList();
+            tutorialIds.Should().NotContain(541); // Tutorial with DiagStatus 0
+            tutorialIds.Should().Equal(535, 536, 537, 538, 539, 540, 542);
+        }
+
+        [Test]
+        public void Get_diagnostic_assessment_should_not_have_archived_tutorials()
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs:202]: Could you somehow mark that this has been done for this page on HEEDLS-288, e.g. by creating a subtask for it and moving it to test or adding a comment? 🙂

File: DigitalLearningSolutions.Data/Services/DiagnosticAssessmentService.cs [hunk 3]
```diff
@@ -64,10 +66,13 @@ FROM Sections
                         Customisations.CustomisationID = @customisationId
                         AND Sections.SectionID = @sectionId
                         AND Sections.ArchivedDate IS NULL
-                        AND Sections.DiagAssessPath IS NOT NULL
+                        AND Sections.DiagAssessPath IS NOT NULL
+                        AND CustomisationTutorials.DiagStatus = 1
+                        AND Tutorials.DiagAssessOutOf > 0
+                        AND Tutorials.ArchivedDate IS NULL
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/DiagnosticAssessmentService.cs:72]: Is the `IsAssessed` check no longer necessary? 🙂

**Review by FridaTveit — COMMENTED** [review]
A few questions, otherwise looks good 🙂

File: DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs (hunk 1 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs:174]: The `DiagStatus` values for the old case were all 0, so we couldn't be sure that the reason the function return `null` was because of the `null` `diagAssessPath`, when it could have just been because of the `DiagStatus` values 🙂

File: DigitalLearningSolutions.Data/Services/DiagnosticAssessmentService.cs (hunk 3 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/DiagnosticAssessmentService.cs:72]: It is not checked by the query given in the ticket, so I removed the check. I'm not entirely sure what `IsAssessed` means, I think it might be talking about post learning assessments rather than diagnostic ones but I'm not to sure 🙂

File: DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs (hunk 2 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs:202]: Done! 🙂

File: DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs (hunk 1 above)
Comment (FridaTveit) [reply DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs:174]: I see, that makes sense 🙂

File: DigitalLearningSolutions.Data/Services/DiagnosticAssessmentService.cs (hunk 3 above)
Comment (FridaTveit) [reply DigitalLearningSolutions.Data/Services/DiagnosticAssessmentService.cs:72]: Yes you're right, I've just checked the ACs for showing the diagnostic assessment on the section page and IsAssessed doesn't need to be true 🙂

**Review by FridaTveit — APPROVED** [review]
Looks good 🙂
